core/new-gui/src/app/workspace/service/workflow-graph/model/shared-model-change-handler.ts

```diff
@@ -0,0 +1,343 @@
+    this.texeraGraph.sharedModel.operatorIDMap.observe((event: Y.YMapEvent<YType<OperatorPredicate>>) => {
+      const jointElementsToAdd: joint.dia.Element[] = [];
+      const newOpIDs: string[] = [];
+      event.changes.keys.forEach((change, key) => {
+        if (change.action === "add") {
+          const newOperator = this.texeraGraph.sharedModel.operatorIDMap.get(key) as YType<OperatorPredicate>;
+          // Also find its position
+          if (this.texeraGraph.sharedModel.elementPositionMap?.has(key)) {
+            const newPos = this.texeraGraph.sharedModel.elementPositionMap?.get(key) as Point;
+            // Add the operator into joint graph
+            const jointOperator = this.jointUIService.getJointOperatorElement(newOperator.toJSON(), newPos);
+            jointElementsToAdd.push(jointOperator);
+            newOpIDs.push(key);
+          } else {
+            throw new Error(`operator with key ${key} does not exist in position map`);
+          }
+        }
+        if (change.action === "delete") {
+          // Disables JointGraph -> TexeraGraph sync temporarily
+          this.texeraGraph.setSyncTexeraGraph(false);
+          // Unhighlight every operator and link to prevent sync error.
+          if (event.transaction.local) {
+            this.jointGraphWrapper.unhighlightElements({
+              operators: this.jointGraphWrapper.getCurrentHighlightedOperatorIDs(),
+              links: this.jointGraphWrapper.getCurrentHighlightedLinkIDs(),
+              groups: [],
+              commentBoxes: [],
+            });
+          }
+          this.jointGraph.getCell(key).remove();
+          // Emit the event streams here, after joint graph is synced.
+          this.texeraGraph.setSyncTexeraGraph(true);
+          this.texeraGraph.operatorDeleteSubject.next({ deletedOperatorID: key });
+        }
+      });
+
+      if (environment.asyncRenderingEnabled) {
+        // Group add
+        this.jointGraphWrapper.jointGraphContext.withContext({ async: true }, () => {
+          this.jointGraph.addCells(jointElementsToAdd);
+        });
+      } else {
+        // Add one by one
+        for (let i = 0; i < jointElementsToAdd.length; i++) {
+          this.jointGraph.addCell(jointElementsToAdd[i]);
+        }
+      }
+
+      // Emit the event streams here, after joint graph is synced and before highlighting.
+      for (let i = 0; i < newOpIDs.length; i++) {
+        const newOpID = newOpIDs[i];
+        const newOperator = this.texeraGraph.sharedModel.operatorIDMap.get(newOpID) as YType<OperatorPredicate>;
+        this.texeraGraph.operatorAddSubject.next(newOperator.toJSON());
+      }
+
+      if (event.transaction.local) {
+        // Only highlight when this is added by current user.
+        this.jointGraphWrapper.setMultiSelectMode(newOpIDs.length > 1);
+        this.jointGraphWrapper.highlightOperators(...newOpIDs);
+      }
+      this.texeraGraph.setSyncTexeraGraph(true);
```

Comment:
There is no symmertic `setSyncTexeraGraph(false)` corresponding to this line.  Line 75 and line 87 is a pair. But this line doesn't have a pair.

I suggest we can remove line 75 and 87, and add `setSyncTexeraGraph(false)` in the very beginning.

Reply:
I believe this line was added by mistake from previous code. Deleted. We need to preserve line 75 and 87 and cannot move `setSyncTexeraGraph(false)` to the beginning because this would result in highlighting bugs. I only turn off `syncTexeraGraph` in these lines to unhighlight when an operator is deleted.